### PR TITLE
FigmaWebhook to MDB cluster

### PR DESCRIPTION
**Pull request by TheSonOfThomp**

### CODE REVIEW

File: components/ContentstackRichText/types.ts [hunk 1]
```diff
@@ -135,7 +135,7 @@ export interface HorizontalLayoutBlockProps {
   column_1: CSNode; // richText
   column_2: CSNode; // richText,
   vertical_align: 'start' | 'center' | 'end' | 'baseline';
-  flex_ratio: `${number}:${number}`
+  flex_ratio: `${number}:${number}`;
```
Comment (brookescarlett) [inline components/ContentstackRichText/types.ts:138]: we should have a linter check like we do with the main repo

File: components/ContentstackRichText/ExampleCardBlock/types.ts [hunk 2]
```diff
@@ -7,6 +7,6 @@ const Variant = {
   Do: 'do',
 } as const;
 
-type Variant = typeof Variant[keyof typeof Variant];
+type Variant = (typeof Variant)[keyof typeof Variant];
```
Comment (brookescarlett) [inline components/ContentstackRichText/ExampleCardBlock/types.ts:10]: not the biggest fan of this rule?

File: package.json [hunk 3]
```diff
@@ -95,6 +95,7 @@
     "eslint-plugin-react": "^7.30.1",
     "eslint-plugin-react-hooks": "^4.6.0",
     "gray-matter": "4.0.3",
+    "mongodb": "^5.1.0",
```
Comment (brookescarlett) [inline package.json:98]: Love to see it 🥇

File: pages/api/figma-publish.ts [hunk 4]
```diff
@@ -1,45 +1,87 @@
...
+import { LibraryPublishEvent } from '../../utils/api/figma.types';
+import { getFigmaVersionHistory } from '../../utils/api/getFigmaVersionHistory';
+import {
+  connectToFigmaVersionsCollection,
+  getLatestEntries,
+  updateFigmaUrl,
+} from '../../utils/api/mdbConnect';
+import { parseUpdatesFromFigmaDescription } from '../../utils/api/parseDescription';
 
 const WEBHOOK_ID = '494792';
```
Comment (brookescarlett) [inline pages/api/figma-publish.ts:14]: is this safe to check in?

File: pages/api/figma-publish.ts [hunk 5]
```diff
@@ -1,45 +1,87 @@
...
-      // 3. Calculate the new version based on the last FigmaVersion and whether `versionUpdate` is a PATCH, MINOR, or MAJOR
-      // 4. POST a new entry to Contentstack with the new version, Component reference, and description
-    });
-  });
-};
-
-export default async function handler(req, res) {
+export default async function handleFigmaPublish(
+  req: {
+    method: 'POST' | 'GET' | 'PUT';
```
Comment (brookescarlett) [inline pages/api/figma-publish.ts:20]: do we not get this type out of the box from Next somewhere?

File: pages/api/figma-publish.ts [hunk 6]
```diff
@@ -1,45 +1,87 @@
...
+
+    const updates = parseUpdatesFromFigmaDescription(body.description);
 
-    if (updatedFile !== FILENAME || requestWebhookId !== WEBHOOK_ID) {
+    if (
+      updatedFile !== FILENAME ||
+      requestWebhookId !== WEBHOOK_ID ||
+      !updates
+    ) {
       res.status(403);
```
Comment (brookescarlett) [inline pages/api/figma-publish.ts:37]: nit: can we send back a better error message?

File: pages/api/figma-publish.ts [hunk 7]
```diff
@@ -1,45 +1,87 @@
...
+    const [currentVersion, prevVersion] = versions;
+    const currVersionUrl = getVersionUrl(currentVersion);
+    const prevVersionUrl = getVersionUrl(prevVersion);
+
+    const { collection, close: closeDB } =
+      await connectToFigmaVersionsCollection();
+    const entries = getLatestEntries({ collection, updates });
+
+    // For each updated component:
+
```
Comment (brookescarlett) [inline pages/api/figma-publish.ts:53]: nit rm space

File: pages/component/[componentName]/changelogs.tsx [hunk 8]
```diff
@@ -145,6 +144,47 @@ const ComponentChangelogs = ({
           ></div>
         )}
       </div>
+      <Button
+        onClick={() => {
+          // TODO: REMOVE THIS
+          const description = '[PATCH] Banner - Rectangle color changed';
```
Comment (brookescarlett) [inline pages/component/[componentName]/changelogs.tsx:150]: intentional?

File: pages/component/[componentName]/changelogs.tsx [hunk 9]
```diff
@@ -167,7 +207,9 @@ export async function getStaticProps({ params: { componentName } }) {
   const figmaChangelog = await getComponentFigmaVersions(
     component ? component.uid : '',
   );
+
   console.log(figmaChangelog[0].component);
```
Comment (brookescarlett) [inline pages/component/[componentName]/changelogs.tsx:211]: rm?

File: tsconfig.json [hunk 10]
```diff
@@ -26,8 +26,8 @@
     "**/*.ts",
     "**/*.tsx",
     "components/**/*.tsx",
-    "utils/*.tsx",
-    "pages/*.tsx"
+    "utils/**/*.ts(x)",
```
Comment (brookescarlett) [inline tsconfig.json:29]: is this new syntax never seen it

File: utils/ContentStack/FigmaVersion.ts [hunk 11]
```diff
@@ -1,20 +1,58 @@
...
+  react_version: string;
+  figma_link: string;
+  updated_at: string;
+}
+
+function parseVersionString(versionString: string): Array<number> | undefined {
+  return versionString.match(/[0-9]+/g)?.map(str => Number(str));
+}
+
+function sortVersions(a: FigmaVersionField, z: FigmaVersionField) {
```
Comment (brookescarlett) [inline utils/ContentStack/FigmaVersion.ts:15]: a? z?

File: utils/MongoDB/connect.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+  Collection,
+  MongoClient,
+  ServerApiVersion,
+} from 'mongodb';
+
+import type {
+  FigmaVersionsMDBDocument,
+} from './mongodb.types';
+
+const uri = `mongodb+srv://${process.env.MDB_USER}:${process.env.MDB_PASSWORD}@figmaversions.fctbuvl.mongodb.net/?retryWrites=true&w=majority`;
```
Comment (brookescarlett) [inline utils/MongoDB/connect.ts:11]: let's not commit this?

File: utils/MongoDB/getLatestEntries.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+export function getLatestEntries({
+  collection,
+  updates,
+}: {
+  collection: Collection<FigmaVersionsMDBDocument>;
+  updates: Array<FigmaComponentUpdate>;
+}) {
+  // Get the latest entries
+  const entries = collection.aggregate<{
```
Comment (brookescarlett) [inline utils/MongoDB/getLatestEntries.ts:18]: ok aggregation pipeline!

File: pages/api/figma-publish.ts (hunk 4 above)
Comment (spark33) [reply pages/api/figma-publish.ts:13]: yeah even the GET request to the webhook requires a Figma token, so this should be safe

**Pull request merged**